src/t_set.c

```diff
@@ -522,24 +728,43 @@ void spopWithCountCommand(client *c) {
      * CASE 2: The number of elements to return is small compared to the
      * set size. We can just extract random elements and return them to
      * the set. */
-    if (remaining*SPOP_MOVE_STRATEGY_MUL > count) {
-        while(count--) {
-            /* Emit and remove. */
-            encoding = setTypeRandomElement(set,&sdsele,&llele);
-            if (encoding == OBJ_ENCODING_INTSET) {
-                addReplyBulkLongLong(c,llele);
-                objele = createStringObjectFromLongLong(llele);
-                set->ptr = intsetRemove(set->ptr,llele,NULL);
+    if (remaining*SPOP_MOVE_STRATEGY_MUL > count &&
+        set->encoding == OBJ_ENCODING_LISTPACK)
+    {
+        /* Specialized case for listpack. Traverse it only once. */
+        unsigned char *lp = set->ptr;
+        unsigned char *p = lpFirst(lp);
+        unsigned int index = 0;
+        while (count) {
+            p = lpNextRandom(lp, p, &index, count--, 0);
```

Comment:
Like here

Reply:
Does it also need to be like other places `serverAssert(p = lpNextRandom(lp, p, &index, count--, 0))`?

Reply:
We already know that count < size and count > 0 here, so it can't return NULL here. Perhaps `assert(count < size)` before the loop?

Assert after every call to listpack functions seems excessive. The we should also assert p != NULL after lpDelete, lpNext, etc. There are already some asserts inside lpNext, etc.